### PR TITLE
[SPARK-47882][SQL] createTableColumnTypes need to be mapped to database types instead of using directly

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

createTableColumnTypes contains Spark SQL data type definitions. The underlying database might not recognize them, boolean for Oracle(v < 23c).

### Why are the changes needed?

bugfix
### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new test

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 1]
```diff
@@ -2182,4 +2183,13 @@ class JDBCSuite extends QueryTest with SharedSparkSession {
     dialect = JdbcDialects.get("jdbc:dummy:dummy_host:dummy_port/dummy_db")
     assert(dialect === NoopDialect)
   }
+
+  test("createTableColumnTypes") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:2187]: Could you add a JIRA ID prefix?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 2]
```diff
@@ -2182,4 +2183,13 @@ class JDBCSuite extends QueryTest with SharedSparkSession {
...
+    val dialect = JdbcDialects.get("jdbc:oracle:dummy_host:dummy_port/dummy_db")
+    val schema = new StructType().add("b", "boolean")
+    val schemaStr =
+      JdbcUtils.schemaString(dialect, schema, caseSensitive = false, Some("b boolean"))
+    assert(schemaStr === """"b" NUMBER(1)""")
+
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:2193]: nit, empty line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 3]
```diff
@@ -875,21 +875,20 @@ object JdbcUtils extends Logging with SQLConfHelper {
...
       .getOrElse(Map.empty[String, String])
-    schema.fields.foreach { field =>
+    schema.foreach { field =>
       val name = dialect.quoteIdentifier(field.name)
       val typ = userSpecifiedColTypesMap
         .getOrElse(field.name, getJdbcType(field.dataType, dialect).databaseTypeDefinition)
-      val nullable = if (field.nullable) "" else "NOT NULL"
-      sb.append(s", $name $typ $nullable")
+      val nullable = if (field.nullable) "" else " NOT NULL"
+      sb.append(s", $name $typ$nullable")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:891]: Please revert this two irrelevant lines.

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun, I have addressed you comments.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you!

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0.

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun

**Pull request closed without merging**